core/src/main/java/org/jruby/RubyBignum.java

```diff
@@ -882,8 +882,21 @@ public IRubyObject op_lshift(ThreadContext context, IRubyObject other) {
         for (;;) {
             if (other instanceof RubyFixnum) {
                 shift = ((RubyFixnum) other).value;
+                if (shift < 0) {
+                    if (value.bitLength() <= -shift ) {
+                        if (value.signum() >= 0) {
+                            return RubyFixnum.zero(context.runtime);
+                        } else {
+                            return RubyFixnum.minus_one(context.runtime);
+                        }
+                    }
+                }
                 break;
             } else if (other instanceof RubyBignum) {
+                if (value == BigInteger.ZERO) {
```

Comment:
think it's safer to not assume only one zero instance:
```suggestion
                if (value.signum() == 0) {
```